src/api/system.c

```diff
@@ -638,6 +638,110 @@ static int f_set_window_opacity(lua_State *L) {
+    P(getmetatable), P(getstack), P(gettable), P(gettop), P(getupvalue),
+    P(getuservalue), P(insert), P(isnumber), 
+    P(isstring), P(isuserdata), P(len), P(load), 
+    P(newstate), P(newthread), P(newuserdata), P(next), 
+    P(pcallk), P(pushboolean), P(pushcclosure), 
+    P(pushfstring), P(pushinteger), P(pushlightuserdata),
+    P(pushlstring), P(pushnil), P(pushnumber), P(pushstring),
+    P(pushthread), P(pushunsigned), P(pushvalue), P(pushvfstring), P(rawequal),
+    P(rawget), P(rawgeti), P(rawgetp), P(rawlen), P(rawset), P(rawseti),
+    P(rawsetp), P(remove), P(replace), P(resume), P(setallocf), 
+    P(setfield), P(setglobal), P(sethook), P(setlocal), P(setmetatable), 
+    P(settable), P(settop), P(setupvalue), P(setuservalue), P(status), 
+    P(tocfunction), P(tointegerx), P(tolstring), P(toboolean),
+    P(tonumberx), P(topointer), P(tothread), 
+    P(tounsignedx), P(touserdata), P(type), P(typename), P(upvalueid),
+    P(upvaluejoin), P(version), P(xmove), P(yieldk)
+  };
+  for (int i = 0; i < sizeof(nodes) / sizeof(lua_function_node); ++i) {
+    if (strcmp(nodes[i].symbol, symbol) == 0)
+      return nodes[i].address;
+  }
+  return NULL;
+}
+
+static int loader_plugin(lua_State* L) {
+  size_t sname, modlen, pathlen;
+  char buffer[512] = "lua_open_";
+  const char* modname = luaL_checklstring(L, -2, &modlen);
+  const char* path = luaL_checklstring(L, -1, &pathlen);
+  void* library = SDL_LoadObject(path);
+  if (modlen == 0 || !library)
+    return luaL_error(L, "Unable to load %s: %s", modname, SDL_GetError());
+  for (sname = modlen - 1; sname > 0 && modname[sname] != '.'; --sname);
+  strcat(buffer, &modname[sname+1]);
+  int (*entrypoint)(lua_State* L, void*) = SDL_LoadFunction(library, buffer);
+  if (!entrypoint) {
+    return luaL_error(L, "Unable to load %s: Can't find entrypoint. Requires a "
+      "function defined as int %s(lua_State* L, void* (*symbol)(const char*))", 
+    modname, buffer);
+  }
+  lua_pushlightuserdata(L, library);
+  lua_setfield(L, LUA_REGISTRYINDEX, modname);
+  if (entrypoint(L, api_require) == 0)
+    return luaL_error(L, "Unable to load %s: Your entrypoint must return at\
+      least one argument.");
+  return 1;
+}
+
+#if _WIN32
+  #define PATH_SEPARATOR '\\'
+  #define DYNAMIC_SUFFIX "dll"
+#else 
+  #if __APPLE__
+    #define DYNAMIC_SUFFIX "lib"
+  #else
+    #define DYNAMIC_SUFFIX "so"
+  #endif
+  #define PATH_SEPARATOR '/'
+#endif
+static int f_searcher_plugin(lua_State *L) {
+  size_t len, cpath_len, cpath_idx = 0, cpath_q = 0;
+  const char* modname = luaL_checklstring(L, -1, &len);
+  lua_getglobal(L, "package");
+  lua_getfield(L, -1, "cpath");
+  const char* cpath = luaL_checklstring(L, -1, &cpath_len);
+  char lib[8192];
+  for (size_t i = 0; i <= cpath_len; ++i) {
+    if (i == cpath_len || cpath[i] == ';') {
+      if (cpath_q) {
+        size_t offset = cpath_q - cpath_idx;
+        strncpy(lib, &cpath[cpath_idx], offset);
```

Comment:
can we be sure that cpath fits in lib?

Reply:
You're right, an additional check should be made here.